django/overwolf_auth/backends.py

```diff
@@ -0,0 +1,147 @@
+
+    # Overwrite values from BaseOAuth2
+    REDIRECT_STATE = False  # Disable state validation
+    # REFRESH_TOKEN_METHOD = "POST"
+    # REFRESH_TOKEN_URL = None
+    # RESPONSE_TYPE = "code"
+    STATE_PARAMETER = False  # Disable state validation
+
+    def extra_data(
+        self,
+        user: Any,
+        uid: Any,
+        response: Response,
+        details: UserDetails = None,
+        *args,
+        **kwargs,
+    ) -> Dict[str, Any]:
+        """
+        Override implementation from BaseOAuth2.
+
+        Augment the info stored in "extra_data" database field with info
+        from the JWT. The JWT itself is also stored, but not needing to
+        decode the token everywhere the info is used makes sense.
+        """
+        data = super().extra_data(user, uid, response, details=details, *args, **kwargs)
+
+        jwt_data = self._decode_response_jwt(response)
+        data["email_verified"] = jwt_data.get("email_verified")
+        data["nickname"] = jwt_data.get("nickname")
+        data["picture"] = jwt_data.get("picture")
+        data["preferred_username"] = jwt_data.get("preferred_username")
+
+        return data
+
+    def get_redirect_uri(self, state: Optional[Any] = None) -> str:
+        """
+        Override implementation from OAuthAuth.
+        """
+        host = (
+            settings.SOCIAL_AUTH_INIT_HOST
+            or settings.AUTH_EXCLUSIVE_HOST
+            or settings.PRIMARY_HOST
+        )
+
+        if host is None:
+            raise ImproperlyConfigured("No host for OverwolfOAuth2")
+
+        path = reverse("social:complete", kwargs={"backend": self.name})
+        return f"{settings.PROTOCOL}{host}{path}"
+
+    def get_user_details(self, response: Response) -> UserDetails:
+        """
+        Override implementation from BaseAuth.
+
+        Returned "know internal struct" shouldn't be changed.
+        """
+        data = self._decode_response_jwt(response)
+
+        try:
+            username = data["sub"]
+        except KeyError:
+            raise AuthException("JWT contained no username (sub)")
+
+        return {
+            "username": username,
+            "email": data.get("email", ""),
+            "fullname": "",
+            "first_name": "",
+            "last_name": "",
+        }
+
+    def get_user_id(self, details: UserDetails, response: Response) -> str:
+        """
+        Override implementation from BaseAuth.
+
+        By default the value is read from response, but it's JWT
+        encoded, so read the value from details instead.
+        """
+        return details[self.ID_KEY]
+
+    def _decode_response_jwt(self, response: Response) -> Dict[str, Any]:
+        """
+        Verify and decode JWT with Overwolf's public key.
+        """
+        token: Optional[str] = response.get("id_token")
+
+        if token is None:
+            raise AuthException("No id_token in auth response")
+
+        jwks_client = PyJWKClient(f"{self.OW_URL}/jwks.json")
```

Comment:
The jwks.json file should probably be cached locally rather than queried every time someone logs in

Reply:
PyJWT's newer versions support caching. Unfortunately they seem to be quite undocumented features. Anyways, [single keys are cached](https://github.com/jpadilla/pyjwt/blob/216cb51e80d36669e235f5fceb5a0fe84e35279c/jwt/jwks_client.py#L40) with [Python's lru_cache](https://docs.python.org/3/library/functools.html#functools.lru_cache) and [key sets](https://github.com/jpadilla/pyjwt/blob/216cb51e80d36669e235f5fceb5a0fe84e35279c/jwt/jwks_client.py#L29) with [timestamps](https://github.com/jpadilla/pyjwt/blob/216cb51e80d36669e235f5fceb5a0fe84e35279c/jwt/jwk_set_cache.py#L7) (that defaults to five minutes). FYI I just tested that login works with the new parameters, I didn't bother to start actually checking what requests the backend sends.

Would this suffice for our needs? If not, do you have suggestions for what approach to take (so it works nicely with our infra)?

Reply:
I'm assuming it's possible to supply the json as an object/file/string rather than an URL? If yes, we should probably use Django's cache backend for the cache, as it's been configured to use redis, which is shared by all instances of the server.

There's a function decorator at `thunderstore.cache.cache.cache_function_result` that should be easy to use for caching the JWKS result (just make sure the function it's applied on returns a JSON serializable type, and not e.g. the full response), though the "cache bust condition" system is a bit outdated (`cache_until` param) and you can simply use `background_update_only` for it.